Fix the invalid amount of rewards distribution

matusvla opened the pull request.

dr497 commented on smart-contract/program/src/processor/crank.rs:
```diff
@@ -61,60 +62,85 @@ pub fn process_crank(
+    msg!("Total staked snapshot {}", central_state.total_staked_snapshot);
+
+    // get the pool staked amount at the time of last system snapshot
+    let total_staked_snapshot: u128 = (stake_pool.header.total_staked - stake_pool.header.total_staked_delta as u64) as u128;
```
You are casting a `i64` into a `u64` this is not safe

dr497 replied:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e00f0da11f00465a0fe161d1cfbf414f

matusvla replied:
Good catch. Could you please check my solution? I am not sure if it is the right one, but I wanted to avoid the overflow of `stake_pool.header.total_staked as i64` as well

dr497 replied:
You are still not solving the issue by casting everything into `i128` and then back into `u128` the issue is that casting signed integer into unsigned integer is unsafe.

Could do something like

```rust
    let total_staked_snapshot = match stake_pool.header.total_staked_delta.signum() {
        1 => stake_pool
            .header
            .total_staked
            .checked_sub(stake_pool.header.total_staked_delta.unsigned_abs()) // <- Actualy safe to cast as u64 here
            .ok_or(AccessError::Overflow)?,
        -1 => stake_pool
            .header
            .total_staked
            .checked_add(stake_pool.header.total_staked_delta.unsigned_abs())
            .ok_or(AccessError::Overflow)?,
        _ => stake_pool.header.total_staked,
    };
```

dr497 replied:
Should also allow you to remove the unnecessary `if total_staked_snapshot > 0`

matusvla replied:
Thank you for the proposed solution, I will use it.

I just want to doublecheck that my solution should never overflow as well, because we can safely assume that `stake_pool.header.total_staked as i128 - stake_pool.header.total_staked_delta as i128 > 0`. Or am I missing something?

As for the  `if total_staked_snapshot > 0`, I believe that we still need it, otherwise we would risk division by 0 on line
` .checked_div(central_state.total_staked_snapshot as u128)`

dr497 replied:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b1f261f46d4ccb75e91d798925d840ec

matusvla replied:
The snippet that you have sent panics on division by 0 which is not something that we want here. We want to save the (0, 0) rewards pair into the rewards history

dr497 commented:
You should run `cargo fmt` this will help format your code

dr497 commented on smart-contract/program/src/processor/claim_bond_rewards.rs:
```diff
@@ -148,8 +144,8 @@ pub fn process_claim_bond_rewards(
-        current_time,
-        bond.last_claimed_time,
+        central_state.last_snapshot_offset,
+        bond.last_claimed_offset as i64,
```
Don't need to cast it it's already a `i64`

dr497 commented on smart-contract/program/src/state.rs:
```diff
@@ -96,6 +93,7 @@ pub struct StakePoolHeader {
     /// Updated by a trustless cranker
     pub current_day_idx: u16,
 
+    // todo maybe pad differently as the fields bellow have changed
```
You don't have to, it's properly aligned

dr497 commented on smart-contract/program/src/state.rs:
```diff
@@ -105,18 +103,17 @@ pub struct StakePoolHeader {
-    // The % of pool rewards going to stakers
-    pub stakers_part: u64,
+    /// Last time the stake pool owner claimed as an offset from the central state's creation time
+    pub last_claimed_offset: i64,
```
If not a timestamp it could be a `u64`

dr497 commented on smart-contract/program/src/state.rs:
```diff
@@ -105,18 +103,17 @@ pub struct StakePoolHeader {
-    /// Last time the stake pool owner claimed
-    pub last_claimed_time: i64,
+    /// The offset of the last total_staked change in this pool from the central state's creation time
+    pub last_delta_update_offset: i64,
```
If not a timestamp it could be a `u64`

dr497 commented on smart-contract/program/src/utils.rs:
```diff
@@ -12,33 +12,34 @@ use spl_token::state::Account;
+        .checked_sub(stake_pool.header.current_day_idx as i64)
         .ok_or(AccessError::Overflow)?
-        > SECONDS_IN_DAY as i64
+        > 0
```
Can simplify by `current_offset > stake_pool.header.current_day_idx`

dr497 commented on smart-contract/program/src/processor/crank.rs:
```diff
@@ -61,60 +62,90 @@ pub fn process_crank(
+        central_state.save(&mut accounts.central_state.data.borrow_mut())?;
+        // reset the delta for the pool that we are currently cranking. We don't want the history to influence our result
+        stake_pool.header.last_delta_update_offset = current_offset;
+        stake_pool.header.total_staked_delta = 0;
```
I think the two lines `L76` and `L77` should be a in a different `if` statement

dr497 replied:
Something like:

```rust
    if stake_pool.header.last_delta_update_offset < current_offset {
        // reset the delta for the pool that we are currently cranking. We don't want the history to influence our result
        stake_pool.header.last_delta_update_offset = current_offset;
        stake_pool.header.total_staked_delta = 0;
    }
```

matusvla replied:
It seems right to me. Could you elaborate on that?

dr497 replied:
If you have 2 pools `P1` and `P2` with `last_snapshot_offset = 0` and `current_offset = 1` but `last_snapshot_offset = 0`

- `P1` gets cranked:
    - `last_snapshot_offset` gets updated to 1
    - `last_delta_update_offset` and `total_staked_delta` get updated
- `P2` gets cranked:
    - `last_snapshot_offset = current_offset`
    - `last_delta_update_offset` and `total_staked_delta` do not get updated
 
Putting this in a different if statement like this would harmonize the crank behavior between pools

```rust
    if stake_pool.header.last_delta_update_offset < current_offset {
        stake_pool.header.last_delta_update_offset = current_offset;
        stake_pool.header.total_staked_delta = 0;
    }
```

matusvla replied:
This unfortunatelly won't work. Imagine the following situation:

- `P1` gets cranked as in your example
- someone stakes into `P2` thereby changing the `total_staked` value recorded for this pool and the value of `total_staked_delta` as well. 
- `P2` gets cranked 
  - We are able to calculate the `total_staked` amount that was in the `P2` at the time of system-wide crank (~`P1` crank in this case)
  - If we set the `total_staked_delta=0` as you're suggesting we wouldn't be able to use it later to derive the value of `total_staked` amount that was in the `P2` at the time of system-wide crank.

dr497 replied:
Isn't it what's already happening with `update_delta` hence why you need to make sure pool are cranked before `stake` and `unstake`?

dr497 replied:
Nvm I think you are right, for readability and maintainability might be easier to harmonize how these fields are updated for all pools in the `crank` instruction and then make sure that all pools are cranked before `stake` and `unstake`.

The result is equivalent to what you currently have but it's just easier to read

matusvla replied:
I don't think that the result is 100% equivalent as this approach doesn't burden the users with cranking. But you are right that it could be hidden in the bindings anyways.

If we just update these fields in the crank method and block the staking for the uncranked pools we probably wouldn't need these fields at all.

Are we sure that forcing the users to crank is a good UX? Because it would happen only very seldom.

dr497 replied:
If the bindings are properly written there is no change to the UX

matusvla replied:
The change has been implemented. Now all the operations that change the pool balance should be blocked when the pool is not cranked.

After discussion with @martincik we have decided that we will keep the bindings simple and atomic and handle the possible need for a crank in our app directly

dr497 commented:
Shouldn't `stake` and `unstake` error if `stake_pool.header.last_delta_update_offset != current_offset` to guarantee everything is always synced?

dr497 commented on smart-contract/program/src/processor/stake.rs:
```diff
@@ -159,14 +159,15 @@ pub fn process_stake(
 
-    let current_time = Clock::get().unwrap().unix_timestamp;
+    let current_time = Clock::get()?.unix_timestamp;
+    let current_offset = (current_time - central_state.creation_time) / SECONDS_IN_DAY as i64;
```
What do you think of moving these two lines into something like this in order to avoid repetition?

```rust
impl CentralState {
    pub fn get_current_offset(&self) -> u64 {
        let current_time = Clock::get().unwrap().unix_timestamp as u64;
        (current_time - self.creation_time as u64) / SECONDS_IN_DAY
    }
}
```

matusvla replied:
Yes, why not :)

dr497 commented on smart-contract/program/src/state.rs:
```diff
@@ -250,36 +243,74 @@ impl StakePoolHeader {
+    }
+
+    // private
+    fn update_delta(
```
I think `total_staked_delta` and `last_delta_update_offset` should be reset in the `crank` instruction. And `stake` and `unstake` should error if the pool and the central state are out of sync.

And then you move this directly into `deposit` and `withdraw`
```rust
        self.total_staked_delta = self
            .total_staked_delta
            .checked_add/sub(amount)
            .ok_or(AccessError::Overflow)?;
        Ok(())
```

matusvla replied:
I disagree with this. With the current implementation it is possible to stake/unstake even when the pool has not been cranked on this day which is a lot less restrictive on the users

dr497 commented on smart-contract/program/src/state.rs:
```diff
@@ -476,15 +511,18 @@ impl CentralState {
             daily_inflation,
             token_mint,
             authority,
+            creation_time: Clock::get()?.unix_timestamp,
```
Given that it will be used as the point of reference for all the program might be more convenient to round it to 00:000 ?

```rust
    pub fn new(
        signer_nonce: u8,
        daily_inflation: u64,
        token_mint: Pubkey,
        authority: Pubkey,
        total_staked: u64,
    ) -> Result<Self, ProgramError> {
        let now = Clock::get()?.unix_timestamp;
        Ok(Self {
            tag: Tag::CentralState,
            signer_nonce,
            daily_inflation,
            token_mint,
            authority,
            creation_time: now - (now - SECONDS_IN_DAY as i64),
            total_staked,
            total_staked_snapshot: 0,
            last_snapshot_offset: 0,
        })
    }
```

matusvla replied:
Could be, but I don't see any benefit in this - time is dependent on the timezone anyways, so it wouldn't be 00:00 in most of the world

dr497 commented on smart-contract/program/src/utils.rs:
```diff
@@ -125,6 +128,11 @@ pub fn assert_uninitialized(account: &AccountInfo) -> ProgramResult {
 
 pub fn assert_authorized_seller(seller: &AccountInfo, seller_index: usize) -> ProgramResult {
+    // print all authorized sellers
+    for i in 0..AUTHORIZED_BOND_SELLERS.len() {
```
`for (i, <item>) in AUTHORIZED_BOND_SELLERS.iter().enumerate()` is a bit more idiomatic (cf `cargo clippy`)

matusvla replied:
Cool stuff, thanks for the tip! Anyways, I have deleted this as it was used only for debugging purposes

matusvla commented:
> Shouldn't `stake` and `unstake` error if `stake_pool.header.last_delta_update_offset != current_offset` to guarantee everything is always synced?

This would be very inconvenient for users as they wouldn't be able to stake/unstake if the pool has not been cranked yet. By introducing the delta we keep track of the `total_staked` changes since the last system-wide crank using the `total_staked_delta` variable and therefore we are able to calculate the rewards appropriately once the pool is cranked.

dr497 commented:
> > Shouldn't `stake` and `unstake` error if `stake_pool.header.last_delta_update_offset != current_offset` to guarantee everything is always synced?
> 
> This would be very inconvenient for users as they wouldn't be able to stake/unstake if the pool has not been cranked yet. By introducing the delta we keep track of the `total_staked` changes since the last system-wide crank using the `total_staked_delta` variable and therefore we are able to calculate the rewards appropriately once the pool is cranked.

Nvm it's actually already the case since users cannot stake/unstake if they have pending rewards and claiming requires the pool to cranked. Only case where it's not already the case is for new stakers

Also, cranking is permissionless so this requirement does not make any difference to the user since this can be abstracted away from them by the bindings

dr497 commented:
> > > Shouldn't `stake` and `unstake` error if `stake_pool.header.last_delta_update_offset != current_offset` to guarantee everything is always synced?
> > 
> > 
> > This would be very inconvenient for users as they wouldn't be able to stake/unstake if the pool has not been cranked yet. By introducing the delta we keep track of the `total_staked` changes since the last system-wide crank using the `total_staked_delta` variable and therefore we are able to calculate the rewards appropriately once the pool is cranked.
> 
> ~Nvm it's actually already the case since users cannot stake/unstake if they have pending rewards and claiming requires the pool to cranked. Only case where it's not already the case is for new stakers~
> 
> Also, cranking is permissionless so this requirement does not make any difference to the user since this can be abstracted away from them by the bindings

If you don't require the pool to be cranked when staking you will allow users to stake + crank + unstake in 1 tx, which in certain circumstances can be net profitable for the user despite the staking fee. Is this intended? If you combine this with a flash loan you could potentially extract a lot of value from the pool without risk.

matusvla commented:
> > > > Shouldn't `stake` and `unstake` error if `stake_pool.header.last_delta_update_offset != current_offset` to guarantee everything is always synced?
> > > 
> > > 
> > > This would be very inconvenient for users as they wouldn't be able to stake/unstake if the pool has not been cranked yet. By introducing the delta we keep track of the `total_staked` changes since the last system-wide crank using the `total_staked_delta` variable and therefore we are able to calculate the rewards appropriately once the pool is cranked.
> > 
> > 
> > ~Nvm it's actually already the case since users cannot stake/unstake if they have pending rewards and claiming requires the pool to cranked. Only case where it's not already the case is for new stakers~
> > Also, cranking is permissionless so this requirement does not make any difference to the user since this can be abstracted away from them by the bindings
> 
> If you don't require the pool to be cranked when staking you will allow users to stake + crank + unstake in 1 tx, which in certain circumstances can be net profitable for the user despite the staking fee. Is this intended? If you combine this with a flash loan you could potentially extract a lot of value from the pool without risk.

Could you please elaborate on this? I don't see how stake + crank + unstake in 1 tx as the rewards are generated only once a day anyways so he cannot collect more rewards by doing this in multiple pools

dr497 commented:
> > > > > Shouldn't `stake` and `unstake` error if `stake_pool.header.last_delta_update_offset != current_offset` to guarantee everything is always synced?
> > > > 
> > > > 
> > > > This would be very inconvenient for users as they wouldn't be able to stake/unstake if the pool has not been cranked yet. By introducing the delta we keep track of the `total_staked` changes since the last system-wide crank using the `total_staked_delta` variable and therefore we are able to calculate the rewards appropriately once the pool is cranked.
> > > 
> > > 
> > > ~Nvm it's actually already the case since users cannot stake/unstake if they have pending rewards and claiming requires the pool to cranked. Only case where it's not already the case is for new stakers~
> > > Also, cranking is permissionless so this requirement does not make any difference to the user since this can be abstracted away from them by the bindings
> > 
> > 
> > If you don't require the pool to be cranked when staking you will allow users to stake + crank + unstake in 1 tx, which in certain circumstances can be net profitable for the user despite the staking fee. Is this intended? If you combine this with a flash loan you could potentially extract a lot of value from the pool without risk.
> 
> Could you please elaborate on this? I don't see how stake + crank + unstake in 1 tx as the rewards are generated only once a day anyways so he cannot collect more rewards by doing this in multiple pools

Actually you would need 2 txs:
- Stake right before crank time
- Crank + claim + unstake

So flash loans would not work but with a normal DeFi loan you could still extract value from the pool without having to actually lock ACCS into the pool. Obviously this would depend on the borrow rate of ACCS, staking fee, total staked and inflation
